### PR TITLE
Simplify connection flow with an asyncio.Protocol

### PR DESCRIPTION
**Pull request by bdraco**
This  removes ~ 20k `call_soon`s per minute which translates to a significant latency improvement (also reduces the run time by ~ 18-24%) with 7 ble proxies.

With this change we can refactor a significant number of coroutines into normal functions.

Eliminates:
- read_loop
- process_loop
- queue

Data now comes in via `data_received` and is called back to the packet processor which was the original `process_loop` and than called back to the consumers of the API (HA) and than to each integration which eliminates the overhead of having to put it into a queue, pull it out of a queue, schedule, fire, etc which was usually more expensive than parsing (when using cpp protobuf) and processing the data itself since almost everything coming from esphome are tiny packets of data (with high frequency with bluetooth active scanning).

### CODE REVIEW

**Comment by bdraco** [comment]
This  removes ~ 20k `call_soon`s per minute which translates to a significant latency improvement (also reduces the run time by ~ 18-24%) with 7 ble proxies

File: aioesphomeapi/model.py [hunk 1]
```diff
@@ -821,20 +822,29 @@ def _convert_bluetooth_le_name(value: bytes) -> str:
...
+            address=data.address,
+            rssi=data.rssi,
+            address_type=data.address_type,
+            name=_convert_bluetooth_le_name(data.name),
+            service_uuids=_convert_bluetooth_le_service_uuids(data.service_uuids),
+            service_data=_convert_bluetooth_le_service_data(data.service_data),
+            manufacturer_data=_convert_bluetooth_le_manufacturer_data(
+                data.manufacturer_data
+            ),
+        )
```
Comment (bdraco) [inline aioesphomeapi/model.py:847]: This belongs in a separate PR

**Comment by bdraco** [comment]
Noticeable improvement with the oral-b toothbrush timer

**Comment by bdraco** [comment]
Need to add noise support and than clean this up

**Comment by bdraco** [comment]
This makes buttons a lot more reliable since they don't stalled anymore

File: aioesphomeapi/_frame_helper.py [hunk 2]
```diff
@@ -35,24 +36,55 @@ class Packet:
     data: bytes
 
 
-class APIFrameHelper(ABC):
+class MissingBytesAPIError(SocketAPIError):
```
Comment (bdraco) [inline aioesphomeapi/_frame_helper.py:39]: Raising an exception for flow control, might not be the best idea here.

File: aioesphomeapi/_frame_helper.py [hunk 3]
```diff
@@ -35,24 +36,55 @@ class Packet:
...
+        """Complete reading a packet from the buffer."""
+        del self._buffer[: self._pos]
+        self._on_pkt(packet)
+
+    def _read_exactly(self, length: int) -> bytes:
+        """Read exactly length bytes from the buffer."""
+        original_pos = self._pos
+        new_pos = original_pos + length
+        if len(self._buffer) < new_pos:
+            raise MissingBytesAPIError(self._buffer, new_pos)
```
Comment (bdraco) [inline aioesphomeapi/_frame_helper.py:84]: Maybe nine instead

File: aioesphomeapi/_frame_helper.py [hunk 4]
```diff
@@ -99,66 +147,68 @@ def write_packet(self, packet: Packet) -> None:
...
-            # If the message is long, we need to read the rest of the length
-            while length[-1] & 0x80 == 0x80:
-                length += await self._reader.readexactly(1)
-
-            # If the message length was longer than 1 byte, we need to read the
-            # message type
-            while not msg_type or (msg_type[-1] & 0x80) == 0x80:
-                msg_type += await self._reader.readexactly(1)
-
-            length_int = bytes_to_varuint(length)
```
Comment (bdraco) [inline aioesphomeapi/_frame_helper.py:143]: This function gets mostly the same input. Might be worth an lru

**Comment by bdraco** [comment]
Working quite well.

Noticeable drop in cpu time

Noticeable decrease in latency controlling ble devices

**Comment by bdraco** [comment]
Reconnects seem to be a bit faster now when I power cycle


```
address_type: 1

2023-01-04 14:01:20.093 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Sending <class 'api_pb2.PingRequest'>: 
2023-01-04 14:01:20.095 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame a0805f307467ede382d486b8291abcbe8574a6b3
2023-01-04 14:01:20.134 INFO (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Ping Failed: Read failed
2023-01-04 14:01:20.135 DEBUG (MainThread) [aioesphomeapi.connection] Cleaning up connection to m5stackatomproxy @ 192.168.106.43
2023-01-04 14:01:20.135 INFO (MainThread) [aioesphomeapi.reconnect_logic] Disconnected from ESPHome API for m5stackatomproxy @ 192.168.106.43
2023-01-04 14:01:20.135 DEBUG (MainThread) [homeassistant.components.esphome] m5stackatomproxy: 192.168.106.43 disconnected, running disconnected callbacks

```

**Comment by bdraco** [comment]
We could probably shave 40-80ms off the reconnect time by sending the hello and login and than waiting for both to come back but I'm not sure its worth it
```
023-01-04 14:01:20.135 INFO (MainThread) [aioesphomeapi.reconnect_logic] Disconnected from ESPHome API for m5stackatomproxy @ 192.168.106.43
2023-01-04 14:01:20.135 DEBUG (MainThread) [homeassistant.components.esphome] m5stackatomproxy: 192.168.106.43 disconnected, running disconnected callbacks
2023-01-04 14:01:20.140 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Connecting to 192.168.106.43:6053 (AddrInfo(family=<AddressFamily.AF_INET: 2>, type=<SocketKind.SOCK_STREAM: 1>, proto=6, sockaddr=IPv4Sockaddr(address='192.168.106.43', port=6053)))
2023-01-04 14:01:20.143 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Opened socket
2023-01-04 14:01:20.144 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame 
2023-01-04 14:01:20.157 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame 00d6bad9492035ee36b606f68a59ce98c3a7b58f83832e939d0507577eeeefa0214d64a65fcfb4f3c8b3dcb3e60e3d1f97
2023-01-04 14:01:20.231 DEBUG (MainThread) [aioesphomeapi._frame_helper] Starting handshake...
2023-01-04 14:01:20.233 DEBUG (MainThread) [aioesphomeapi._frame_helper] Handshake complete
2023-01-04 14:01:20.233 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Sending <class 'api_pb2.HelloRequest'>: client_info: "Home Assistant 2023.2.0.dev0"
api_version_major: 1
api_version_minor: 7

2023-01-04 14:01:20.233 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame 689a53e169ee31d21dd13b1e3be925667caec8f1e26bebfb62ee2f3949dab99e988079b859dbba8aecbcf6365d2407713a60e1ee5a56
2023-01-04 14:01:20.246 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Got message of type <class 'api_pb2.HelloResponse'>: api_version_major: 1
api_version_minor: 7
server_info: "m5stackatomproxy (esphome v2023.1.0-dev)"
name: "m5stackatomproxy"

2023-01-04 14:01:20.246 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Successfully connected ('m5stackatomproxy (esphome v2023.1.0-dev)' API=1.7)
2023-01-04 14:01:20.246 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Sending <class 'api_pb2.ConnectRequest'>: 
2023-01-04 14:01:20.247 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame d98f700f9d0b62f4c858973fdbbd75b2158d7782
2023-01-04 14:01:20.265 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Got message of type <class 'api_pb2.GetTimeRequest'>: 
2023-01-04 14:01:20.265 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Sending <class 'api_pb2.GetTimeResponse'>: epoch_seconds: 1672876880

2023-01-04 14:01:20.265 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame bfe51f3a1702ec90da7a06b25dbb7c38f0ae734f818ed7541a
2023-01-04 14:01:20.266 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Got message of type <class 'api_pb2.ConnectResponse'>: 
2023-01-04 14:01:20.266 INFO (MainThread) [aioesphomeapi.reconnect_logic] Successfully connected to m5stackatomproxy @ 192.168.106.43
2023-01-04 14:01:20.267 DEBUG (MainThread) [aioesphomeapi.connection] 192.168.106.43: Sending <class 'api_pb2.DeviceInfoRequest'>: 
2023-01-04 14:01:20.267 DEBUG (MainThread) [aioesphomeapi._frame_helper] Sending frame 8badaf9698e8d4369008c7966a703736ac4ac4b3
2023-01-04 14:01:20.330 DEBUG (MainThread) [aioesphomeapi.connection] m5stackatomproxy @ 192.168.106.43: Got message of type <class 'api_pb2.DeviceInfoResponse'>: name: "m5stackatomproxy"
mac_address: "E8:9F:6D:0A:68:E0"
esphome_version: "2023.1.0-dev"
compilation_time: "Jan  3 2023, 17:17:56"
model: "m5stack-atom"
bluetooth_proxy_version: 3
manufacturer: "Espressif"


```

File: aioesphomeapi/connection.py [hunk 5]
```diff
@@ -112,10 +111,6 @@ def __init__(
 
         self._ping_stop_event = asyncio.Event()
 
-        self._to_process: asyncio.Queue[Optional[Packet]] = asyncio.Queue()
-
-        self._process_task: Optional[asyncio.Task[None]] = None
```
Comment (balloob) [inline aioesphomeapi/connection.py:117]: 🔥

File: aioesphomeapi/connection.py [hunk 6]
```diff
@@ -545,115 +525,83 @@ async def send_message_await_response(
...
-                    pkt.data,
-                    e,
-                    exc_info=True,
-                )
-                await self._report_fatal_error(
-                    ProtocolAPIError(f"Invalid protobuf message: {e}")
-                )
-                raise
+        self._handle_fatal_error(err)
+        asyncio.create_task(self._cleanup())
```
Comment (balloob) [inline aioesphomeapi/connection.py:545]: `self._cleanup` doesn't do any await and could also be made non-async.

**Review by balloob — APPROVED** [review]
Awesome 😎 🐬

File: aioesphomeapi/connection.py (hunk 6 above)
Comment (bdraco) [reply aioesphomeapi/connection.py:545]: Good idea. I'll do a follow up to address that

**Comment by bdraco** [comment]
Thanks

File: aioesphomeapi/connection.py (hunk 6 above)
Comment (bdraco) [reply aioesphomeapi/connection.py:545]: Started a PR here https://github.com/esphome/aioesphomeapi/pull/355 

Its a bit more involved that I was expecting but should still finish it up today
